File: task4/task4_python/Player.py
import random
import abc
from Pos import Pos
from Weapon import Weapon
class Player(object):
    __metaclass__ = abc.ABCMeta
    def __init__(self, healthCap, mob, posx, posy, index, game):
        self.__MOBILITY = mob
        self._health = healthCap
        self._pos = Pos(posx, posy)
        self._index = index
        self._game = game
        self._myString = ""
        self._equipment = ""
    
    def getPos(self):
        return self._pos
    
    def teleport(self):
        randx = random.randint(0, self._game.D - 1)
        randy = random.randint(0, self._game.D - 1)
        while(self._game.positionOccupied(randx, randy)):
            randx = random.randint(0, self._game.D - 1)
            randy = random.randint(0, self._game.D - 1)
        self._pos.setPos(randx, randy)
    
    def increaseHealth(self, h):
        self._health = h + self._health
    
    def decreaseHealth(self, h):
        self._health = self._health - h
        if (self._health <= 0):
            self._myString = "C" + self._myString[0]
    
    def getName(self):
        return self._myString
    
    def askForMove(self):
        print "Your health is ", 
        print self._health ,
        print ". Your position is ({0},{1}). Your mobility is {2}.".format(self._pos.getX(), self._pos.getY(), self.__MOBILITY)
        print "You now have following options: "
        print "1. Move"
        print "2. Attack"
        print "3. End the turn"

        a = int(raw_input())

        if (a == 1):
            print "Specify your target position (Input 'x y')."
            posx = int(raw_input())
            posy = int(raw_input())
            if(self._pos.distance(posx, posy) > self.__MOBILITY):
                print "Beyond your reach. Staying still."
            elif self._game.positionOccupied(posx, posy):
                print "Position occupied. Cannot move there."
            else:
                self._pos.setPos(posx, posy)
                self._game.printBoard()
                print "You can now "
                print "1.attack"
                print "2.End the turn"
                if (int(raw_input()) == 1):
                    print "Input position to attack. (Input 'x y')"
                    attx = int(raw_input())
                    atty = int(raw_input())
                    self._equipment.action(attx, atty)
        elif (a == 2):
            print "Input position to attack."
            attx = int(raw_input())
            atty = int(raw_input())
            self._equipment.action(attx, atty)
        elif (a == 3):
            return

File: task4/task4_java/Player.java
import java.util.Random;

public abstract class Player {
	private int MOBILITY;
	protected Pos pos;
	protected int health;
	protected Weapon equipment;
	protected int index;
	protected String myString;
	protected SurvivalGame game;
	
	public Player(int healthCap, int mob, int posx, int posy, int index, SurvivalGame game) {

		this.MOBILITY = mob;
		this.health = healthCap;
		this.pos = new Pos(posx, posy);
		this.index = index;
		this.game = game;
	}

	public Pos getPos() {
		return pos;
	}

	public void teleport() {

		Random rand;
		rand = new Random();
		int randx = rand.nextInt(game.D);
		int randy = rand.nextInt(game.D);
		while (game.positionOccupied(randx, randy)) {
			randx = rand.nextInt(game.D);
			randy = rand.nextInt(game.D);
		}
		pos.setPos(randx, randy);
	}

	public void increaseHealth(int h) {
		this.health += h;
		if (this.health > 0){
			if (this instanceof Chark){
				this.myString = "C" + Integer.toString(this.index);
			}
			else{
				this.myString = "H" + Integer.toString(this.index);
			}
		}
	}

	public void decreaseHealth(int h) {
		this.health -= h;
		if (this.health <= 0)
			this.myString = "C" + this.myString.charAt(0);
	}

	public String getName() {
		return myString;
	}

	public void askForMove() {
		// Print general information
		System.out.println("Your health is " + health
				+ String.format(". Your position is (%d,%d). Your mobility is %d.", pos.getX(), pos.getY(), this.MOBILITY));
		
		if (this.equipment instanceof Axe || this.equipment instanceof Rifle) {

			System.out.println("You now have following options: ");
			System.out.println("1. Move");
			System.out.println("2. Attack");
			System.out.println("3. End tne turn");

			int a = SurvivalGame.reader.nextInt();

			if (a == 1) {
				System.out.println("Specify your target position (Input 'x y').");
				int posx = SurvivalGame.reader.nextInt(), posy = SurvivalGame.reader.nextInt();
				if (pos.distance(posx, posy) > this.MOBILITY) {
					System.out.println("Beyond your reach. Staying still.");
				} else if (game.positionOccupied(posx, posy)) {
					System.out.println("Position occupied. Cannot move there.");
				} else {
					this.pos.setPos(posx, posy);
					game.printBoard();
					System.out.println("You can now \n1.Attack \n2.End the turn");
					if (SurvivalGame.reader.nextInt() == 1) {
						System.out.println("Input position to attack. (Input 'x y')");
						int attx = SurvivalGame.reader.nextInt(), atty = SurvivalGame.reader.nextInt();
						this.equipment.action(attx, atty);
					}
				}
			} else if (a == 2) {
				System.out.println("Input position to attack.");
				int attx = SurvivalGame.reader.nextInt(), atty = SurvivalGame.reader.nextInt();
				this.equipment.action(attx, atty);
			} else if (a == 3) {
				return;
			}
		}
		else if (this.equipment instanceof Wand){
			System.out.println("You now have following options: ");
			System.out.println("1. Move");
			System.out.println("2. Heal");
			System.out.println("3. End tne turn");

			int a = SurvivalGame.reader.nextInt();
			if (a == 1) {
				System.out.println("Specify your target position (Input 'x y').");
				int posx = SurvivalGame.reader.nextInt(), posy = SurvivalGame.reader.nextInt();
				if (pos.distance(posx, posy) > this.MOBILITY) {
					System.out.println("Beyond your reach. Staying still.");
				} else if (game.positionOccupied(posx, posy)) {
					System.out.println("Position occupied. Cannot move there.");
				} else {
					this.pos.setPos(posx, posy);
					game.printBoard();
					System.out.println("You can now \n1.Heal\n2.End the turn");
					if (SurvivalGame.reader.nextInt() == 1) {
						System.out.println("Input position to heal. (Input 'x y')");
						int attx = SurvivalGame.reader.nextInt(), atty = SurvivalGame.reader.nextInt();
						this.equipment.action(attx, atty);
					}
				}
			} else if (a == 2) {
				System.out.println("Input position to heal.");
				int attx = SurvivalGame.reader.nextInt(), atty = SurvivalGame.reader.nextInt();
				this.equipment.action(attx, atty);
			} else if (a == 3) {
				return;
			}
		}
	}

}
